## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [ ]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [6]:
# Write your code for importing the csv file here
import numpy as np
import pandas as pd
movies = pd.read_csv('/Users/hailing/Downloads/6_Pandas_Exercise_2_Movie_Dataset/6_Movies.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [25]:
import numpy as np
import pandas as pd
df = pd.read_csv('/Users/hailing/Downloads/6_Pandas_Exercise_2_Movie_Dataset/6_Movies.csv')

# Inspect columns
print(df.columns)

# Inspect shape
print(df.shape)

# Inspect variable types
print(df.dtypes)



Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')
(3853, 28)
color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
m

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [16]:
import numpy as np
import pandas as pd
df = pd.read_csv('/Users/hailing/Downloads/6_Pandas_Exercise_2_Movie_Dataset/6_Movies.csv')

print(df.shape)


(3853, 28)


#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [24]:
null_column = df.isnull().any()
sum_null_column = null_column.sum()

print(sum_null_column)


12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [29]:
import numpy as np
import pandas as pd
df = pd.read_csv('/Users/hailing/Downloads/6_Pandas_Exercise_2_Movie_Dataset/6_Movies.csv')
newdf=df.drop(columns=['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio','plot_keywords'],axis=1)
print(newdf)


          director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Ch

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [23]:
newdf=df.drop(columns=['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio','plot_keywords'],axis=1)
print(newdf.shape[1])


13


**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [21]:
newdf = df.drop(columns=['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'], axis=1)

null_percentage = (newdf.isnull().sum() / len(newdf)) * 100

formatted_percentage = null_percentage.apply(lambda x: f"{x:.2f}%")

print(formatted_percentage)


director_name             0.00%
num_critic_for_reviews    0.03%
gross                     0.00%
genres                    0.00%
actor_1_name              0.00%
movie_title               0.00%
num_voted_users           0.00%
num_user_for_reviews      0.00%
language                  0.08%
budget                    0.00%
title_year                0.00%
imdb_score                0.00%
movie_facebook_likes      0.00%
dtype: object


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [22]:
newdf = df.drop(columns=['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'], axis=1)

null_percentage = (newdf.isnull().sum() / len(newdf)) * 100

formatted_percentage = null_percentage.apply(lambda x: f"{x:.2f}%")

highest_null_column = null_percentage.idxmax()
print(highest_null_column)

language


**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [19]:
newdf['language'] = newdf['language'].replace(np.nan, 'English')

print(newdf['language'])


0       English
1       English
2       English
3       English
4       English
         ...   
3848    English
3849    English
3850    Spanish
3851    English
3852    English
Name: language, Length: 3853, dtype: object


#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

In [372]:
eng_df = newdf[newdf.language == 'English']
print(len(eng_df)+1)




3672


### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [150]:
eng_df = df[df.language == 'English']
eng_df = df[df.language == 'English'].copy()
eng_df['budget'] = eng_df['budget'] / 1000000
eng_df['gross'] = eng_df['gross'] / 1000000

print(eng_df['budget'], eng_df['gross'])

0       237.0000
1       300.0000
2       245.0000
3       250.0000
4       263.7000
          ...   
3844      0.0100
3848      0.0070
3849      0.0070
3851      0.0090
3852      0.0011
Name: budget, Length: 3671, dtype: float64 0       760.505847
1       309.404152
2       200.074175
3       448.130642
4        73.058679
           ...    
3844      0.180483
3848      0.424760
3849      0.070071
3851      0.004584
3852      0.085222
Name: gross, Length: 3671, dtype: float64


**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [71]:
eng_df['profit'] = eng_df['gross'] - eng_df['budget']



In [354]:
eng_df = eng_df.sort_values(by='profit', ascending=False)


In [363]:
top10 = eng_df.head(10)
print(top10)



      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
28    Color    Colin Trevorrow                   644.0     124.0   
25    Color      James Cameron                   315.0     194.0   
2704  Color       George Lucas                   282.0     125.0   
2748  Color   Steven Spielberg                   215.0     120.0   
16    Color        Joss Whedon                   703.0     173.0   
482   Color       Roger Allers                   186.0      73.0   
230   Color       George Lucas                   320.0     136.0   
64    Color  Christopher Nolan                   645.0     152.0   
419   Color          Gary Ross                   673.0     142.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
28                      365.0                  1000.0         Judy Greer   
25                     

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [369]:
print(top10[4:5])

2748    E.T. the Extra-Terrestrial 
Name: movie_title, dtype: object


**Subtask 3.3: Find IMDb Top 200**

Create a new dataframe `IMDb_Top_200` and store the top 200 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 200 indicating the ranks of the corresponding films.

In [103]:
IMDb_Top_200 = eng_df[eng_df.num_voted_users >= 25000].sort_values('imdb_score', ascending=False).head(200)
IMDb_Top_200['Rank'] = range(1, 201)

print(IMDb_Top_200)

      color         director_name  num_critic_for_reviews  duration  \
1795  Color        Frank Darabont                   199.0     142.0   
3016  Color  Francis Ford Coppola                   208.0     175.0   
2543  Color  Francis Ford Coppola                   149.0     220.0   
64    Color     Christopher Nolan                   645.0     152.0   
325   Color         Peter Jackson                   328.0     192.0   
...     ...                   ...                     ...       ...   
3237  Color          Duncan Jones                   415.0      97.0   
668   Color     Quentin Tarantino                   596.0     187.0   
3405  Color     Richard Linklater                   405.0     109.0   
1734  Color          Pierre Morel                   309.0      93.0   
1502  Color       Nick Cassavetes                   177.0     123.0   

      director_facebook_likes  actor_3_facebook_likes          actor_2_name  \
1795                      0.0                   461.0        Jeffrey

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_200*? </font>

In [133]:
bucket_1 = len(IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 7.5) & (IMDb_Top_200['imdb_score'] < 8)])
bucket_2 = len(IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 8) & (IMDb_Top_200['imdb_score'] < 8.5)])
bucket_3 = len(IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 8.5) & (IMDb_Top_200['imdb_score'] < 9)])
bucket_4 = len(IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 9) & (IMDb_Top_200['imdb_score'] < 9.5)])
bucket_5 = len(IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 9.5) & (IMDb_Top_200['imdb_score'] < 10)])

buckets={'7.5 to 8':bucket_1,'8 to 8.5':bucket_2,'8.5 to 9':bucket_3,'9 to 9.5':bucket_4,'9.5 to 10':bucket_5}

print(max(buckets,key=buckets.get))


8 to 8.5


**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [277]:
Meryl_Streep=eng_df[eng_df.actor_1_name=='Meryl Streep']

In [16]:
Leo_Caprio=eng_df[eng_df.actor_1_name=='Leonardo DiCaprio']


In [151]:
Brad_Pitt=eng_df[eng_df.actor_1_name=='Brad Pitt']

In [159]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt])


In [160]:
grouped_combined = Combined.groupby('actor_1_name')


In [340]:

mean_critic_reviews = grouped_combined['num_critic_for_reviews'].mean()
mean_user_reviews = grouped_combined['num_user_for_reviews'].mean()

highest_mean_critic = mean_critic_reviews.idxmax()
highest_mean_user = mean_user_reviews.idxmax()

print("Actor with the highest mean num_critic_for_reviews:", highest_mean_critic)
print("Actor with the highest mean num_user_for_reviews:", highest_mean_user)


Actor with the highest mean num_critic_for_reviews: Leonardo DiCaprio
Actor with the highest mean num_user_for_reviews: Leonardo DiCaprio


In [342]:
mean_critic = round(grouped_combined['num_critic_for_reviews'].mean(), 2)
mean_audience = round(grouped_combined['num_user_for_reviews'].mean(), 2)

print('Actor with the highest mean of critic reviews:', mean_critic.idxmax(), max(mean_critic))
print('Actor with the highest mean of audience reviews:', mean_audience.idxmax(),max(mean_audience))


Actor with the highest mean of critic reviews: Leonardo DiCaprio 330.19
Actor with the highest mean of audience reviews: Leonardo DiCaprio 914.48


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [329]:
highest_rated1 = round(grouped_combined['num_critic_for_reviews'].max(), 2)

print( highest_rated1.idxmax())

Leonardo DiCaprio


#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [339]:
highest_rated = round(grouped_combined['num_user_for_reviews'].max(), 2)

print( highest_rated.idxmax())

Brad Pitt


#### <font color='red'>Question 10: Which movie starting with the letter A, has the 3rd highest profit?</font>
-  <font color='red'>Austin Powers: The Spy Who Shagged Me</font>
-  <font color='red'>Avatar</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [253]:
eng_df['profit'] = eng_df['gross'] - eng_df['budget']
filtered_movies = eng_df[eng_df['movie_title'].str.startswith('A', na=False)]
eng_df1 = filtered_movies.sort_values(by='profit', ascending=False)

print(eng_df1.iloc[2]['movie_title'])

Avengers: Age of Ultron 


#### <font color='red'>Question 11: Among the movies starting with the letter A, who has the highest imdb score?</font>
-  <font color='red'>American History X</font>
-  <font color='red'>Apocalypse Now</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [252]:
eng_df1 = filtered_movies.sort_values(by='imdb_score', ascending=False)
highest_imdb = eng_df1.iloc[0]['movie_title']

print(highest_imdb)


American History X 


#### <font color='red'>Question 12: Among the movies that '*Joss Whedon*' has directed, which of them made the highest profit?</font>
-  <font color='red'>Serenity</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [257]:
joss_whedon = eng_df[eng_df.director_name=='Joss Whedon']

sort_by_profit= joss_whedon.sort_values(by='profit', ascending=False)
highest_profit = sort_by_profit.iloc[0]['movie_title']

print(highest_profit)


Avengers: Age of Ultron 


#### <font color='red'>Question 13: Among the movies that were released in the year '*2007*' & '*2008*', which of them had the 2nd highest imdb score?</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>WALL·E</font>
-  <font color='red'>No End in Sight</font>
-  <font color='red'>Into the Wild</font>

In [328]:
year_2007=eng_df[eng_df.title_year==2007.0]
year_2008=eng_df[eng_df.title_year==2008.0]
Combined = pd.concat([year_2007,year_2008])

eng_df2 = Combined.sort_values(by='imdb_score', ascending=False)

print(eng_df2.iloc[1:2]['movie_title'])


56    WALL·E 
Name: movie_title, dtype: object
